### PR TITLE
[DeviceMesh] Initialized mesh tensor with CPU context

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124787
* #124780
* #124768
* __->__ #124767
* #124741
* #124651

This PR makes sure to construct the `DeviceMesh`'s `mesh` tensor on CPU device in `init_device_mesh()`. This means that we can call `init_device_mesh()` under meta-device context and still construct the correct `mesh` tensor.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/device_mesh.py [hunk 1]
```diff
@@ -553,7 +553,10 @@ def init_device_mesh(
                     f"Found len(mesh_dim_names): {len(mesh_dim_names)} and len(mesh_shape):{len(mesh_shape)}.",
                 )
 
-        mesh = torch.arange(math.prod(mesh_shape), dtype=torch.int).view(mesh_shape)
+        # Always initialize the mesh's tensor on CPU, regardless of what the
+        # external device type has been set to be (e.g. meta)
+        with torch.device("cpu"):
```
Comment (wz337) [inline torch/distributed/device_mesh.py:558]: Oh I think with this, we can remove the .detach().to("cpu") here: 
https://github.com/pytorch/pytorch/blob/main/torch/distributed/device_mesh.py#L215

Comment (awgu) [reply torch/distributed/device_mesh.py:558]: I was worrying that someone out there might init a `DeviceMesh` object directly instead of through `init_device_mesh`, so leaving that there might be safer.

Comment (wz337) [reply torch/distributed/device_mesh.py:558]: Just remember we had a check right above it, but ye I think we can keep it to make sure the dtype is consistent. 
https://github.com/pytorch/pytorch/blob/main/torch/distributed/device_mesh.py#L212-L213

**Review by wz337 — APPROVED** [review]
LGTM
